src/net.c

```diff
@@ -1354,6 +1428,95 @@ coap_read(coap_context_t *ctx, coap_tick_t now) {
+ * directly saves having to iterate through the endpoints / sessions.
+ */
+void
+coap_io_do_events(coap_context_t *ctx, struct epoll_event *events, int nevents) {
+  coap_tick_t now;
+  int j;
+
+  coap_ticks(&now);
+  for(j = 0; j < nevents; j++) {
+    coap_socket_t *sock = (coap_socket_t*)events[j].data.ptr;
+
+    /* Ignore 'timer trigger' ptr  which is NULL */
+    if (sock) {
+      if (sock->endpoint) {
+        coap_endpoint_t *endpoint = sock->endpoint;
+        if ((sock->flags & COAP_SOCKET_WANT_READ) &&
+            (events[j].events & EPOLLIN)) {
+          sock->flags |= COAP_SOCKET_CAN_READ;
+          coap_read_endpoint(endpoint->context, endpoint, now);
+        }
+
+        if ((sock->flags & COAP_SOCKET_WANT_WRITE) && (events[j].events & EPOLLOUT)) {
+          /*
+           * Need to update this to EPOLLIN as EPOLLOUT will normally always
+           * be true causing epoll_wait to return early
+           */
+          coap_epoll_ctl_mod(sock, EPOLLIN, __FUNCTION__);
+          sock->flags |= COAP_SOCKET_CAN_WRITE;
+          coap_write_endpoint(endpoint->context, endpoint, now);
+        }
+
+        if ((sock->flags & COAP_SOCKET_WANT_ACCEPT) &&
+            (events[j].events & EPOLLIN)) {
+          sock->flags |= COAP_SOCKET_CAN_ACCEPT;
+          coap_accept_endpoint(endpoint->context, endpoint, now);
+        }
+
+      }
+      else if (sock->session) {
+        coap_session_t *session = sock->session;
+
+        if ((sock->flags & COAP_SOCKET_WANT_CONNECT) &&
+            (events[j].events & (EPOLLOUT|EPOLLERR|EPOLLHUP|EPOLLRDHUP))) {
+          sock->flags |= COAP_SOCKET_CAN_CONNECT;
+          /* Make sure the session object is not deleted in one of the callbacks  */
+          coap_session_reference(session);
+          coap_connect_session(session->context, session, now);
+          coap_session_release(session);
+        }
+
+        if ((sock->flags & COAP_SOCKET_WANT_READ) &&
+            (events[j].events & EPOLLIN)) {
+          sock->flags |= COAP_SOCKET_CAN_READ;
+          /* Make sure the session object is not deleted in one of the callbacks  */
+          coap_session_reference(session);
+          coap_read_session(session->context, session, now);
+          coap_session_release(session);
+        }
+
+        if ((sock->flags & COAP_SOCKET_WANT_WRITE) && (events[j].events & EPOLLOUT)) {
+          /*
+           * Need to update this to EPOLLIN as EPOLLOUT will normally always
+           * be true causing epoll_wait to return early
+           */
+          coap_epoll_ctl_mod(sock, EPOLLIN, __FUNCTION__);
+          sock->flags |= COAP_SOCKET_CAN_WRITE;
+          /* Make sure the session object is not deleted in one of the callbacks  */
+          coap_session_reference(session);
+          coap_write_session(session->context, session, now);
+          coap_session_release(session);
+        }
+      }
+    }
+#ifdef HAVE_SYS_TIMERFD_H
+    else if (ctx->eptimerfd != -1) {
```

Comment:
@mrdeep1 I'm confused about the code snippet. It seemed that it does nothing to do with coap?

Reply:
coap_io_do_events() is a replacement for coap_read (which works on the file descriptors), and so is a vital part of keeping libcoap's I/O running when coap_run_once() is called (following a suggestion earlier in this PR and saves having to loop through all the endpoints / sessions).

Are you running into some kind of issue with the epoll code that you are trying to resolve?

Reply:
No, no. I just wonder what's the timer fd and read it here. I found it only read it but did nothing then.

Reply:
The purpose of the read is to clear the event created on eptimerfd.  If the read does not take place, then the next epoll_wait() will immediately return with the event set.

Reply:
Is it still OK to disable HAVE_SYS_TIMERFD_H but enable EPOLL?

Reply:
Interesting question.  I am glad that we are getting a good code review!

If we have code of the form
```
while (1) {
  ...
  coap_run_once(ctx, some_value_gt_1);
  ...
}
```
there will be no issues as coap_run_once() will get repeatedly called.  However if the code is of the form where the exposed FD is being monitored for FD activity along with other FDs and coap_run_once() is only called if there is activity on the exposed FD, then any packet re-transmissions etc. will not take place with timerfd not being available until there is any new incoming packet associated with libcoap.  Not an ideal solution.

As both epoll and timerfd have been in the Linux kernel for some time (over 10 years), both should be available, but timerfd arrived later in kernel version 2.6.22.

If timerfd is not available, it needs to be replaced with something different - perhaps a pipe with a timer process at the other end of the threat that can be primed to write something to the pipe whenever the next re-transmit etc. needs to take place.

I will put in some warning code that both EPOLL and TIMERFD are required.